packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -224,14 +228,18 @@ export const LightningElement: LightningElementConstructor = function (
 
     // Linking elm, shadow root and component with the VM.
     associateVM(component, vm);
-    associateVM(cmpRoot, vm);
+    if (shouldAttachShadow) {
```

Comment:
Instead of checking if the `shouldAttachShadow` and retrieving the shadow root from `vm.cmpRoot`, I would recommend hoisting the shadow root outside if the if block and running `associateVM` and `patchShadowRootWithRestrictions` if the shadow root is not undefined. It improves overall typings and avoid unnecessary property lookup.


```ts
let shadowRoot: ShadowRoot | undefined
if (hasShadow(VM) {
  shadowRoot = vm.cmpRoot = renderer.attachShadow( /* ... */);
}

// ...

if (!isUndefined(shadowRoot)) {
  associateVM(vm.cmpRoot, VM);
}

// ...

if (!isUndefined(shadowRoot)) {
  patchShadowRootWithRestrictions(shadowRoot);
}
```